File: 18-4sum/18-4sum.py
class Solution:
  def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
    l = len(nums)
    if l < 4: return []
    if l == 4 and sum(nums) == target: return [nums]
    nums.sort()
    avg = target / 4
    if nums[0] >= avg + 1: return []
    if nums[-1] < avg: return []
    
    result = []
    a = 0
    while a < l-3:
      na = nums[a]
      b = a + 1
      while b < l-2:
        nb = nums[b]
        rest = target - na - nb
        c = b + 1
        d = l - 1
        while c < d:
          nc = nums[c]
          nd = nums[d]
          s = nc + nd
          if s < rest:
            while c < d and nums[c] == nc: c += 1
          elif s == rest:
            result += [[na, nb, nc, nd]]
            while c < d and nums[c] == nc: c += 1
            while c < d and nums[d] == nd: d -= 1
          else:
            while c < d and nums[d] == nd: d -= 1
        while b < l-2 and nums[b] == nb: b += 1
      while a < l-3 and nums[a] == na: a += 1
    return result

File: 18-4sum/18-4sum.go
func fourSum(nums []int, target int) [][]int {
  length := len(nums)
  if length < 4 { return [][]int{} }
  avg := int(math.Ceil(float64(target >> 2)))
  avg2 := int(math.Ceil(float64(target >> 1)))
  sort.Ints(nums)
  ans := make([][]int, 0, length - 2)
  ai := nums[0] - 1
  for i := 0; i < length - 3; i++ {
    for ; i < length - 3 && nums[i] == ai; i++ {}
    if i >= length - 3 { break }
    ai = nums[i]
    if ai > avg { break }
    aj := nums[i+1] - 1
    for j := i + 1; j < length - 2; j++ {
      for ; j < length - 2 && nums[j] == aj; j++ {}
      if j >= length - 2 { break }
      aj = nums[j]
      sum := ai + aj
      if sum > avg2 { break }
      t := target - sum
      l, r := j+1, length - 1
      for l < r {
        al := nums[l]
        ar := nums[r]
        if al + ar == t {
          ans = append(ans, []int{ai, aj, al, ar})
          for l++; l < r && nums[l] == al; l++ {}
          for r--; l < r && nums[r] == ar; r-- {}
        } else if al + ar > t {
          for r--; l < r && nums[r] == ar; r-- {}
        } else {
          for l++; l < r && nums[l] == al; l++ {}
        }
      }
    }
  }
  return ans
}